js/ui/tag_box.js

```diff
@@ -850,30 +850,39 @@ var TagBox = SelectBox.inherit({
             filteredValues[filteredItemValue] = filteredItem;
         }.bind(this));
 
+        var loadItemPromises = [];
+
         values.forEach(function(value, index) {
             var currentItem = filteredValues[isValueExprSpecified ? JSON.stringify(value) : value];
 
             if(isValueExprSpecified && !isDefined(currentItem)) {
-                this._loadItem(value, cache).always((function(item) {
-                    this._createTagData(items, item, value, index);
-                }).bind(this));
+                loadItemPromises.push(this._loadItem(value, cache).always((function(item) {
+                    var newItem = this._createTagData(items, item, value, index);
+                    items.splice(index, 0, newItem);
+                }).bind(this)));
             } else {
-                this._createTagData(items, currentItem, value, index);
+                var newItem = this._createTagData(items, currentItem, value, index);
+                items.splice(index, 0, newItem);
             }
         }.bind(this));
 
-        return items;
+        var d = new Deferred();
```

Comment:
```suggestion
return $.when.apply(this, loadItemPromises); 
```

Reply:
Deferred should be resolved with pre-filled array of items. We can not use this construction